mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -291,6 +358,11 @@ LAGenericRewritePattern::matchAndRewrite(linalg::GenericOp laGeneric,
     return failure();
   }
 
+  if (isLeadingToBlockArg(gemmStoreOp.getDest().getDefiningOp())) {
```

Comment:
> From what I can tell, this code will always clone the threadwise_write_all

Why would you think that ? 

It is meant clone if the gemmStoreOp dest is leading to output buffer -- which is not always case. In the single output fusion case, the prior to aligntiling the output of gemmStoreOp dest is a memref that is an input to linalg block. Hence, we dont need to clone that case because it is not a real output of the kernel.

However, if that output was mem copied to an output buffer, then we need to clone (instead of modifying the gemmStoreOp).

Do you want me to put a description on the code ?

Reply:
If what you are saying is true -- we'd have three threadwise_write_all ops (1 original and 2 for two linalg blocks)

Reply:
Ok, wait, hold on, I thought `gemmStoreOp` was the `threadwise_write_all`

Reply:
Because I thought we had an infinite loop here: find the store op, clone it, find that, clone that, and so on

Reply:
(I ll add unit tests once we agree on the changes)

Reply:
The thing is, the threadwise_write_all gets modified to point at the generic output after fusion. So I could see this code leading to having extra threadwise_write_all's

Reply:
> The thing is, the threadwise_write_all gets modified to point at the generic output after fusion. So I could see this code leading to having extra threadwise_write_all's 

No it get modifed to point to internal memref alloca that copy optimized later if it is actually an output

Reply:
(These internal memref allocas are useful as they are bit SSA'ish)

Reply:
"the generic output" means the output of the linalg.generic. And then the rewrite for copies come is and the threadwise_write_all gets updated again to the output of the memref.copy, and so on

Reply:
I'm basically wanting to make sure this code is clear on future reading. Currently, it wasn't. Maybe it's variable naming.

Reply:
> Because I thought we had an infinite loop here: find the store op, clone it, find that, clone that, and so on

In any case, it wouldn't be an infinite loop; it would overclone max per linalg block (per rewrite).

But to your point, yes Im relying on the internal alloc s and their copies distinguish which are outputs and which are internal based on the fact what gets copied out to output memrefs because it is the state of IR we receive today. 

I do acknowledge the point if a better bufferization pass will eliminate those internal allocas -- we might need to bit more analysis to identify outputs vs internal states but IMO we should not run that until later -- if possible.

Reply:
And, I'll note, that in the single argument case, the output leads to a block argument too? And so it would be cloned?

Reply:
In the single output case, the linalg block will have two memref : input and output.

The rock op will be writing to the input memref of linalg (which is not copied, hence not cloned)

Reply:
(I've added a description now -- PTAL)

Reply:
I ll also add unit test as promised before, once we agree on the changes -- which would also make this clearer....